File: Arrays and String/498. Diagonal Traverse/traverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        diag = []
        i = 0
        j = 0
        n = len(mat)
        m = len(mat[0])
        adder = -1

        diag.append(mat[i][j])

        while i < n - 1 or j < m - 1:
            if j < m - 1 and ((i == 0 and adder == -1) or (i == n - 1 and adder == 1)):
                j += 1
                adder = -adder
            elif (j == 0 and adder == 1) or (j == m - 1 and adder == -1):
                i += 1
                adder = -adder
            else:
                i += adder
                j -= adder

            diag.append(mat[i][j])

        return diag

File: 3. Arrays and String/498. Diagonal Traverse/traverse.cpp
class Solution
{
public:
    vector<int> findDiagonalOrder(vector<vector<int>> &mat)
    {
        vector<int> diag;
        int i = 0;
        int j = 0;
        int n = mat.size();
        int m = mat[0].size();
        int adder = -1;

        diag.push_back(mat[i][j]);

        while (i < n - 1 || j < m - 1)
        {
            if (j < m - 1 && ((i == 0 && adder == -1) || (i == n - 1 && adder == 1)))
            {
                ++j;
                adder = -adder;
            }
            else if ((j == 0 && adder == 1) || (j == m - 1 && adder == -1))
            {
                ++i;
                adder = -adder;
            }
            else
            {
                i += adder;
                j -= adder;
            }
            diag.push_back(mat[i][j]);
        }
        return diag;
    }
};